src/Bicep.Core/TypeSystem/TypeFactory.cs

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System.Collections.Concurrent;
+
+namespace Bicep.Core.TypeSystem;
+
+public static class TypeFactory
+{
+    private record struct IntegerAttributes(long? MinValue, long? MaxValue, TypeSymbolValidationFlags ValidationFlags);
+    private record struct IntegerLiteralAttributes(long Value, TypeSymbolValidationFlags ValidationFlags);
+    private record struct ArrayAttributes(long? MinLength, long? MaxLength, TypeSymbolValidationFlags ValidationFlags);
+    private record struct StringAttributes(long? MinLength, long? MaxLength, TypeSymbolValidationFlags ValidationFlags);
+    private record struct StringLiteralAttributes(string Value, TypeSymbolValidationFlags ValidationFlags);
+
+    private static ConcurrentDictionary<IntegerAttributes, IntegerType> IntegerTypePool = new();
+    private static ConcurrentDictionary<IntegerLiteralAttributes, IntegerLiteralType> IntegerLiteralTypePool = new();
+    private static ConcurrentDictionary<ArrayAttributes, ArrayType> ArrayTypePool = new();
+    private static ConcurrentDictionary<StringAttributes, StringType> StringTypePool = new();
+    private static ConcurrentDictionary<StringLiteralAttributes, StringLiteralType> StringLiteralTypePool = new();
```

Comment:
I'm not seeing much value in these caches from the benchmarks. I think the LanguageServer benchmark is where the effect would be most noticeable, but enabling the caches seems to reduce allocations by ~1%, reduce Gen0 garbage collections by ~0.5%, and increase runtime by ~3%:

### With caches:

``` ini

BenchmarkDotNet=v0.13.4, OS=Windows 11 (10.0.22621.1344)
Intel Core i7-1065G7 CPU 1.30GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK=7.0.103
  [Host]     : .NET 7.0.3 (7.0.323.6910), X64 RyuJIT AVX2
  DefaultJob : .NET 7.0.3 (7.0.323.6910), X64 RyuJIT AVX2


```
|                                 Method |    Mean |    Error |   StdDev |        Gen0 |       Gen1 | Allocated |
|--------------------------------------- |--------:|---------:|---------:|------------:|-----------:|----------:|
| &#39;Type a document from start to finish&#39; | 4.745 s | 0.1907 s | 0.5380 s | 209000.0000 | 23000.0000 |    1.2 GB |

### Without caches: 

``` ini

BenchmarkDotNet=v0.13.4, OS=Windows 11 (10.0.22621.1344)
Intel Core i7-1065G7 CPU 1.30GHz, 1 CPU, 8 logical and 4 physical cores
.NET SDK=7.0.103
  [Host]     : .NET 7.0.3 (7.0.323.6910), X64 RyuJIT AVX2
  DefaultJob : .NET 7.0.3 (7.0.323.6910), X64 RyuJIT AVX2


```
|                                 Method |    Mean |    Error |   StdDev |        Gen0 |       Gen1 | Allocated |
|--------------------------------------- |--------:|---------:|---------:|------------:|-----------:|----------:|
| &#39;Type a document from start to finish&#39; | 4.605 s | 0.1244 s | 0.3487 s | 210000.0000 | 23000.0000 |   1.21 GB |

Overall, I think any benefit would need to be substantiated by more than these results. At this point, I don't think it's worth the additional complexity, though I'll leave the type factory in place so that we can introduce caching in the future if desired.